File: android/src/main/java/org/maplibre/reactnative/components/camera/MLRNCameraManager.java
package org.maplibre.reactnative.components.camera;

import org.maplibre.geojson.FeatureCollection;
import com.facebook.common.logging.FLog;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;
import org.maplibre.reactnative.components.AbstractEventEmitter;
import org.maplibre.reactnative.utils.GeoJSONUtils;

import java.util.HashMap;
import java.util.Map;

public class MLRNCameraManager extends AbstractEventEmitter<MLRNCamera> {
    public static final String REACT_CLASS = "MLRNCamera";

    private ReactApplicationContext mContext;

    public MLRNCameraManager(ReactApplicationContext context) {
        super(context);
        mContext = context;
    }

    @Override
    public Map<String, String> customEvents() {
        return new HashMap<>();
    }

    @Override
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    protected MLRNCamera createViewInstance(ThemedReactContext reactContext) {
        return new MLRNCamera(reactContext, this);
    }

    @ReactProp(name="stop")
    public void setStop(MLRNCamera camera, ReadableMap map) {
        if (map != null) {
            CameraStop stop = CameraStop.fromReadableMap(mContext, map, null);
            camera.setStop(stop);
        }
    }

    @ReactProp(name="defaultStop")
    public void setDefaultStop(MLRNCamera camera, ReadableMap map) {
        if (map != null) {
            CameraStop stop = CameraStop.fromReadableMap(mContext, map, null);
            camera.setDefaultStop(stop);
        }
    }

    @ReactProp(name="maxBounds")
    public void setMaxBounds(MLRNCamera camera, String value) {
        if (value != null) {
            FeatureCollection collection = FeatureCollection.fromJson(value);
            camera.setMaxBounds(GeoJSONUtils.toLatLngBounds(collection));
        }
    }


    @ReactProp(name="userTrackingMode")
    public void setUserTrackingMode(MLRNCamera camera, int userTrackingMode) {
        camera.setUserTrackingMode(userTrackingMode);
        throw new AssertionError("Unused code");
    }

    @ReactProp(name="followZoomLevel")
    public void setZoomLevel(MLRNCamera camera, double zoomLevel) {
        camera.setZoomLevel(zoomLevel);
    }

    @ReactProp(name="followUserLocation")
    public void setFollowUserLocation(MLRNCamera camera, boolean value) {
        camera.setFollowUserLocation(value);
    }

    @ReactProp(name="followUserMode")
    public void setFollowUserMode(MLRNCamera camera, String value) {
        camera.setFollowUserMode(value);
    }

    @ReactProp(name="minZoomLevel")
    public void setMinZoomLevel(MLRNCamera camera, double value) {
        camera.setMinZoomLevel(value);
    }

    @ReactProp(name="maxZoomLevel")
    public void setMaxZoomLevel(MLRNCamera camera, double value) {
        camera.setMaxZoomLevel(value);
    }

    @ReactProp(name="followPitch")
    public void setFollowPitch(MLRNCamera camera, double value) {
        camera.setFollowPitch(value);
    }

}

File: android/src/main/java/org/maplibre/reactnative/components/camera/MLRNCameraManager.kt
package org.maplibre.reactnative.components.camera

import com.facebook.react.bridge.Dynamic
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.MLRNCameraManagerDelegate
import com.facebook.react.viewmanagers.MLRNCameraManagerInterface
import org.maplibre.geojson.FeatureCollection
import org.maplibre.reactnative.components.AbstractEventEmitter
import org.maplibre.reactnative.utils.GeoJSONUtils


@ReactModule(name = MLRNCameraManager.REACT_CLASS)
class MLRNCameraManager(private val mContext: ReactApplicationContext) :
    AbstractEventEmitter<MLRNCamera>(
        mContext
    ), MLRNCameraManagerInterface<MLRNCamera> {
    private val delegate: MLRNCameraManagerDelegate<MLRNCamera, MLRNCameraManager> =
        MLRNCameraManagerDelegate(this)

    override fun getDelegate(): ViewManagerDelegate<MLRNCamera> = delegate

    companion object {
        const val REACT_CLASS: String = "MLRNCamera"
    }

    override fun getName(): String = REACT_CLASS


    override fun customEvents(): Map<String, String> {
        return HashMap()
    }

    override fun createViewInstance(reactContext: ThemedReactContext): MLRNCamera {
        return MLRNCamera(reactContext, this)
    }

    @ReactProp(name = "stop")
    override fun setStop(camera: MLRNCamera, map: Dynamic) {
        if (!map.isNull) {
            camera.setStop(map.asMap())
        }
    }

    @ReactProp(name = "defaultStop")
    override fun setDefaultStop(camera: MLRNCamera, map: Dynamic) {
        if (!map.isNull) {
            val stop = CameraStop.fromReadableMap(mContext, map.asMap(), null)
            camera.setDefaultStop(stop)
        }
    }

    @ReactProp(name = "maxBounds")
    override fun setMaxBounds(camera: MLRNCamera, value: String?) {
        if (value != null) {
            val collection = FeatureCollection.fromJson(value)
            camera.setMaxBounds(GeoJSONUtils.toLatLngBounds(collection))
        }
    }

    @ReactProp(name = "userTrackingMode")
    override fun setUserTrackingMode(camera: MLRNCamera, userTrackingMode: Int) {
        camera.setUserTrackingMode(userTrackingMode)
        throw AssertionError("Unused code")
    }

    @ReactProp(name = "followZoomLevel")
    override fun setFollowZoomLevel(camera: MLRNCamera, zoomLevel: Double) {
        camera.setZoomLevel(zoomLevel)
    }

    @ReactProp(name = "followUserLocation")
    override fun setFollowUserLocation(camera: MLRNCamera, value: Boolean) {
        camera.setFollowUserLocation(value)
    }

    @ReactProp(name = "followUserMode")
    override fun setFollowUserMode(camera: MLRNCamera, value: String?) {
        camera.setFollowUserMode(value)
    }

    override fun setFollowHeading(camera: MLRNCamera, value: Double) {
        // TODO: Not implemented?
    }

    override fun setFollowPadding(camera: MLRNCamera, value: Dynamic) {
        // TODO: Not implemented?
    }

    @ReactProp(name = "followPitch")
    override fun setFollowPitch(camera: MLRNCamera, value: Double) {
        camera.setFollowPitch(value)
    }

    @ReactProp(name = "zoomLevel")
    override fun setZoomLevel(camera: MLRNCamera, zoomLevel: Double) {
        camera.setZoomLevel(zoomLevel)
    }

    @ReactProp(name = "minZoomLevel")
    override fun setMinZoomLevel(camera: MLRNCamera, value: Double) {
        camera.setMinZoomLevel(value)
    }

    @ReactProp(name = "maxZoomLevel")
    override fun setMaxZoomLevel(camera: MLRNCamera, value: Double) {
        camera.setMaxZoomLevel(value)
    }

    override fun setAnimationDuration(camera: MLRNCamera, value: Double) {
        // iOS only, no-op on Android
    }

    override fun setAnimationMode(camera: MLRNCamera, value: String?) {
        // iOS only, no-op on Android
    }
}
